src/eval_utils/run.ts

```diff
@@ -0,0 +1,725 @@
+                    JSON.stringify(file.version) +
+                    ", got: " +
+                    JSON.stringify(file.callable.version),
+            );
+        }
+    }
+
+    let type: FileType;
+    if (file.type) {
+        type = file.type;
+        console.info(
+            `${CYAN}Evaluating your ${type} function corresponding to ${file.path} on Humanloop${RESET}\n\n`,
+        );
+    } else {
+        type = "flow";
+        console.warn("No file type specified, defaulting to flow.");
+    }
+
+    const function_ = file.callable;
+    if (!function_) {
+        if (type === "flow") {
+            throw new Error(
+                "You must provide a callable for your Flow file to run a local eval.",
+            );
+        } else {
+            console.info(
+                `No callable provided for your ${type} file - will attempt to generate logs on Humanloop.`,
+            );
+        }
+    }
+
+    const { callable, version, ...rest } = file;
+    let hlFile: FileResponse;
+    switch (type) {
+        case "flow": {
+            // Be more lenient with Flow versions as they are arbitrary json
+            if (version && !version.attributes) {
+                version.attributes = version as Record<string, unknown>;
+            }
+            const updatedData = { ...rest, ...version } as FlowRequest;
+            hlFile = await client.flows.upsert(updatedData);
+            break;
+        }
+        case "prompt": {
+            hlFile = await client.prompts.upsert({
+                ...rest,
+                ...version,
+            } as PromptRequest);
+            break;
+        }
+        case "tool": {
+            hlFile = await client.tools.upsert({
+                ...rest,
+                ...version,
+            } as ToolRequest);
+            break;
+        }
+        case "evaluator": {
+            // @ts-ignore EvaluatorRequest is generated by Fern as 'unknown'
+            // Leading to a type error here
+            hlFile = await client.evaluators.upsert({
+                ...rest,
+                ...version,
+            } as EvaluatorsRequest);
+            break;
+        }
+        default:
+            throw new Error(`Unsupported File type: ${type}`);
+    }
+
+    // Upsert the dataset
+    if (dataset.action === undefined) {
+        dataset.action = "set";
+    }
+    if (dataset.datapoints === undefined) {
+        //  Use `upsert` to get existing dataset ID if no datapoints provided, given we can't `get` on path.
+        dataset.datapoints = [];
+        dataset.action = "set";
+    }
+    let hlDataset = await client.datasets.upsert(dataset);
+    hlDataset = await client.datasets.get(hlDataset.id, {
+        versionId: hlDataset.versionId,
+        includeDatapoints: true,
+    });
+
+    // Upsert the local Evaluators; other Evaluators are just referenced by `path` or `id`
+    let localEvaluators: [EvaluatorResponse, Function][] = [];
```

Comment:
is our terminology here "local evaluators"?
in Humanloop we call them "External Evaluators".
is this something taken from the python sdk?

Reply:
Yes - should I change to externalEvaluators?

Reply:
yes to which question?

Reply:
Yes it's taken from python SDK. I think it's good enough